tests/python/pants_test/rules/test_run.py

```diff
@@ -0,0 +1,56 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.build_graph.address import Address, BuildFileAddress
+from pants.engine.fs import Digest, FileContent, InputFilesContent, Workspace
+from pants.engine.interactive_runner import InteractiveRunner
+from pants.rules.core import run
+from pants.rules.core.binary import CreatedBinary
+from pants_test.console_rule_test_base import ConsoleRuleTestBase
+from pants_test.engine.util import MockConsole, run_rule
+
+
+class RunTest(ConsoleRuleTestBase):
+  goal_cls = run.Run
+
+  def create_mock_binary(self, program_text: bytes) -> CreatedBinary:
+    input_files_content = InputFilesContent((
+      FileContent(path='program.py', content=program_text, is_executable=True),
+    ))
+    digest, = self.scheduler.product_request(Digest, [input_files_content])
+    return CreatedBinary(
+      command='program.py',
+      digest=digest,
+    )
+
+  def single_target_run(self, console: MockConsole, program_text: bytes, spec: str):
```

Comment:
Nit: I recommend requiring keyword only args. None of these are intuitive as positional only args.

Then, on line 44, I'd personally rewrite it to this:

```python
res = self.single_target_run(
  console=MockConsole(use_colors=False), 
  program_text=b'#!/usr/bin/python\nprint("hello")', 
  spec="some/addr"
)
```

Reply:
Also, `spec` and `console` are the same for both tests. Can we remove them from being parameters and hardcode in this util function?

Reply:
the point of passing `console` into the function is so that the actual test method has access to it after `single_target_run` completes, so it can query the stdout/stderr on it.

Reply:
The reason I had the spec be an argument to the util function was because the text of the spec is part of the test, so I wanted to have them defined near each other.